Request-Reply sample code

HariRangarajan-Solace opened the pull request:
- Implemented code for blocking request/reply pattern

Commit:
Implemented code for blocking request/reply pattern

Mrc0113 commented on src/main/java/com/solace/samples/java/patterns/DirectReplierBlocking.java:
```diff
@@ -0,0 +1,156 @@
+
+        //5. Build and start the Receiver object
+        RequestReplyMessageReceiver requestReplyMessageReceiver = messagingService.requestReply().
+                createRequestReplyMessageReceiverBuilder().build(TopicSubscription.of(TOPIC_PREFIX + "*/direct/>"));
```
We should add `request` in the topic so this replier only tries to respond to requests. Looks like @TamimiGitHub used `solace/samples/python/direct/request/>` for the Python replier.

HariRangarajan-Solace replied:
Done

Mrc0113 commented on src/main/java/com/solace/samples/java/patterns/DirectReplierBlocking.java:
```diff
@@ -0,0 +1,156 @@
+ * <p>
+ * Refer to the DirectRequesterBlocking class for the request component of the flow.
+ */
+public class DirectReplierBlocking {
```
Should this be just `DirectReplier`. I don't think there is anything in here that makes it `Blocking` at first glance.

Mrc0113 commented on src/main/java/com/solace/samples/java/patterns/DirectReplierBlocking.java:
```diff
@@ -0,0 +1,156 @@
+            //This SOP is just for demo purposes, ideally considering the slow nature of console I/O, any such action should be avoided in message processing
+            System.out.println("The outbound message is : " + outboundMessage.getPayloadAsString());
+
+            //7-D. Post the reply to the incoming message
```
For your own knowledge take a look at how the utility method handles the message IDs, does it take the request's message ID and add it as the correlation ID in the response? or does it do something else? I believe there is also a reply flag that gets set to make it work with the classic APIs ;) 

**Consider extending the comment to explain how this works if you think it isn't too much info. If it's more than a few lines then let's just make sure it's captured in the tutorial that @aaron-613 started

HariRangarajan-Solace replied:
So I checked it out and indeed it generates a sequential identifier for each request : **#REQ0**, **#REQ1** and so on for each message that is published by the requestor. 
This id is used in the **CorrelationId** header.
The sequential identifier is specific to each session or execution.

HariRangarajan-Solace replied:
Also, the explanation regarding the whole behaviour is very well explained over here : https://tutorials.solace.dev/jcsmp/request-reply/ and I have put a link to this page in both the class comments.

Mrc0113 commented on src/main/java/com/solace/samples/java/patterns/DirectRequesterBlocking.java:
```diff
@@ -0,0 +1,156 @@
+                //10. Define the topic name where the message will be posted.
+                // This example defines a dynamic topic for each message to showcase Solace capability of wildcard based topics
+                // In cases where dynamic topics names are not required, a simple topic string can be used
+                final String topicString = new StringBuilder(TOPIC_PREFIX).append(API.toLowerCase()).append("/direct/pub/").append(chosenCharacter).toString();  // StringBuilder faster than +
```
Similar to comment on Replier, let's put `request` in the topic

Mrc0113 commented:
At first glance I think this looks pretty good besides the couple of comments I made but I'll leave it to @aaron-613 to do a more thorough review.

Commit:
Minor code refactorings and topic changes

aaron-613 commented:
Hi Hari!  I wish we had a chat before you started writing these samples as I could have pointed you in the right direction.  :-/

Ok, first biggest comment: I think you should start with the two Direct request/reply samples from JCSMP as a starting point, and copy/update code to make it work with Java API... easier than trying to modify the Direct Pub/Sub samples to look like the request/reply ones.  The payloads are different than the publisher/subscriber samples, and some other stuff.

If don't wish to rewrite, then at first glance here are comments to make look similar to JCMSP req/reply samples...

**Requestor** sample:
- Remove unused `APPROX_MSG_RATE_PER_SEC` and `PAYLOAD_SIZE` static vars
- Don't populate payload with random chars but use similar payload to JCSMP
- Don't need message counters, again look at JCSMP req/reply samples.  Get rid of `registerLoggerForMonitoring()`
- I prefer the code self-contained in one larger method, rather than calling small little helper methods that are only used once.  Easier to read the code straight-through than jumping down for validation and registering.
    - If helper methods were called lots of times, then yes definitely.
- Not sure why, but in output of Replier app, it's saying the Correlation ID is a GUID, but I can see it is not.  Actually, there appears to be a correlation ID in the user property map too??
```
Destination:                            Topic 'solace/samples/java/direct/request/H'
CorrelationId:                          #REQ33                  <---- ?
Priority:                               4
Class Of Service:                       USER_COS_1
DeliveryMode:                           DIRECT
Message Id:                             38
ReplyTo:                                Topic '#P2P/v:solace913/kWgl05tr/app_3feffcfec21705ba/_'
DMQ Eligible
User Property Map:                      2 entries
  Key 'solace.messaging.message.correlation-id' (String): f74df284-9df5-4866-a3b7-f71e511783ce   <----- ?
  Key 'solace.messaging.message.application-message-id' (String): f74df284-9df5-4866-a3b7-f71e511783ce
```
- Not sure which is being used?  `#REQ33` is the JCSMP correlation ID.  Ah, found it:
- 
```
        final String header_id = UUID.randomUUID().toString();
        messageBuilder.withProperty(SolaceProperties.MessageProperties.APPLICATION_MESSAGE_ID, header_id);  // as an example of a header
        messageBuilder.withProperty(SolaceProperties.MessageProperties.CORRELATION_ID, header_id);  // as an example of a header
        return messageBuilder.build(payloadString);  // binary payload message
```
- Ok, these are User Properties... not actual message headers.  These are unnecessary.  (should be anyhow).

On the **Replier** code:
- Lots of similar comments as above
- Get rid of discard checking, and message rate reporting
- This actually is BLOCKING sample, from your line here: `requestReplyMessageReceiver.receiveMessage(messageHandler);`
     - Is that the best way to do it?  The JCSMP uses a callback handler.  This style is different than I'm used to.  Let's discuss with @Mrc0113 
     - That is why the while loop is not printing out message received rates every second, it's blocked inside the while loop, and then Sleeps for 1 second after.


That's it for now.  I think it'd be easier to take the JCSMP samples and massage them into Java API samples.  But if not, then have both open side-by-side and change these samples to look more like those.

Good luck!

aaron-613 commented:
Oh and add run scripts to the `build.gradle` file.  So you can compile/run from the command line..!?
```
    'DirectRequesterBlocking':'com.solace.samples.java.patterns.DirectRequesterBlocking',
    'DirectReplierBlocking':'com.solace.samples.java.patterns.DirectReplierBlocking',
```

Just doing some random testing on the command line, if the Requestor is started first before Replier, it never recovers.

Commit:
Changes for fixing review comments

TamimiGitHub commented on src/main/java/com/solace/samples/java/patterns/DirectReplierBlocking.java:
```diff
@@ -0,0 +1,143 @@
+
+            //This SOP is just for demo purposes, ideally considering the slow nature of console I/O, any such action should be avoided in message processing
+//            System.out.println("The inbound message is : " + inboundMessage.dump()); // Enable this for learning purposes as it logs a String representation of the whole Message
+            System.out.println("The inbound message payload is : " + inboundMessage.getPayloadAsString());
```
This will fail if the payload is not string. You can test it out by sending a message using the broker management console and you can specify the type of message as text or bytes.

One way to avoid this is to check the type of the payload and use the appropriate method to extract the payload. For example, this is how it's done in python
https://github.com/SolaceSamples/solace-samples-python/blob/master/patterns/direct_subscriber.py#L18-L22
and this is how we do it in go
https://github.com/SolaceSamples/solace-samples-go/blob/main/patterns/direct_subscriber.go#L18-L25

You can also check out @aaron-613's comment on this community post 
https://solace.community/discussion/comment/5259#Comment_5259

HariRangarajan-Solace replied:
Hey @TamimiGitHub  : nice catch. I forgot the point that our payload are transmitted and processed as bytes and we dont have any restriction on the payload that can be published to a topic. 
But I found a nice api method which helps in this.

[InboundMessage. getAndConvertPayload](https://docs.solace.com/API-Developer-Online-Ref-Documentation/pubsubplus-java/com/solace/messaging/receiver/InboundMessage.html#getAndConvertPayload-com.solace.messaging.util.Converter.BytesToObject-java.lang.Class-) which we can use to supply a converter and this is responsible for converting from the byte[] to any format that you want to work with. In this case I have stuck to String but made a comment on its flexibility.

`final String stringPayload = inboundMessage.getAndConvertPayload(
                    (byte[] bytes) -> {
                        return new String(bytes);
                    }, String.class
            );`

Hope this fixes and addressess your point.

TamimiGitHub replied:
ouuuh was not aware of [getAndConvertPayload](https://docs.solace.com/API-Developer-Online-Ref-Documentation/pubsubplus-java/com/solace/messaging/receiver/InboundMessage.html#getAndConvertPayload-com.solace.messaging.util.Converter.BytesToObject-java.lang.Class-)

Looks like this will be a good alternative. @aaron-613 is there something similar in JCSMP?

Commit:
Changes for fixing review comments

HariRangarajan-Solace commented:
> Oh and add run scripts to the `build.gradle` file. So you can compile/run from the command line..!?
> 
> ```
>     'DirectRequesterBlocking':'com.solace.samples.java.patterns.DirectRequesterBlocking',
>     'DirectReplierBlocking':'com.solace.samples.java.patterns.DirectReplierBlocking',
> ```
> 
> Just doing some random testing on the command line, if the Requestor is started first before Replier, it never recovers.

@aaron-613 : i updated the gradle file. 
For the failing requester, IMO its sufficient behaviour. Since we are expecting a reply to an outbound message, if one does not come in within the timeout period the producer stops. Now this is not production quality as ideally you would want the producer to either continue publishing (and loosing messages without any value) or handle the situation more gracefully like some metrices or errors etc. I have made a note of it in the exception handling block.

aaron-613 commented:
Just finished review.  I previously suggested looking at JCSMP req/reply samples to see how they did things, and would recommend again.  They are different than the pub/sub samples, topics, timeouts, functionality exposed, etc.  Additionally, interoperability between the two should be tested... we are striving to be able to mix-and-match any API with any other.  For example: the topics do not line up here between JCSMP and Java: `solace/samples/jcsmp/direct/request` vs. `solace/samples/java/direct/request/43`.  Please test Java<->Java, JCSMP<->Java, and Java<->JCSMP.

Ok, comments:

**Requestor:**
- Filename, class name, and run scripts should be "requestOr", not "requestEr"
- `REQUEST_TIMEOUT_MS = 10000;` seems a bit long?  Default in JCSMP and others is 3000
- I'm still on the fence about these "helper" methods that only have a few lines in them, since you have to jump around in the code to see what they're doing.  Also, this is one is just 1 line, so why not have in the main body?
```
    private static String createMessagePayload(final int loopCounter) {
        // each loop, change the payload, less trivial
        return String.format("Hello, this is the request number:  #%d", loopCounter);    <--- extra space too
    }
```
- The ApplicationMessageID is still not an actual app msg ID, it's a User Property.  But this is an API problem, not your code.  I'll ping the R&D guys about it.  I have the same issue on the DirectPublisher.  Here's the diff between Java API and JCSMP API:
```
JCSMP DirectPublisher:
^^^^^^^^^^^^^^^^^^ Start Message ^^^^^^^^^^^^^^^^^^^^^^^^^^^
Destination:                            Topic 'solace/samples/jcsmp/direct/pub/V'
ApplicationMessageId:                   790c72d5-2fc0-4a44-9211-e1789626687e
SequenceNumber:                         74

Java DirectPublisher:
^^^^^^^^^^^^^^^^^^ Start Message ^^^^^^^^^^^^^^^^^^^^^^^^^^^
Destination:                            Topic 'solace/samples/java/direct/pub/V'
DMQ Eligible
User Property Map:
  Key 'solace.messaging.message.application-message-id' (STRING) 3d1eed99-5a03-44d4-ad91-5f26ea2c3372
```
- I'd suggest removing it from the Request/Reply samples as it's not super necessary here and already in the publisher.  And we're going to have to correct it once R&D fixes the API.
- The Requestor errors out and quits if it doesn't receive a response in the time allotted.  Perhaps make it retry, loop, like JCSMP?
```
 } catch (JCSMPRequestTimeoutException e) {
      System.out.println("Failed to receive a reply in " + REQUEST_TIMEOUT_MS + " msecs");
  } catch (JCSMPException e) {  // threw from send(), only thing that is throwing here, but keep trying (unless shutdown?)
      System.out.printf("### Caught while trying to producer.send(): %s%n", e);
```
- Means you have `catch` for exactly the right exception (line 87) since everything in Java API is a RuntimeException now.  Also, the error message on the console is incorrect: `Caught while trying to publisher.publish():` should be `publishAwaitResponse()`

**Replier:**
- Was this code run before submitting??  I'm getting unhandled exceptions the first time I try to run.
```
--LOG-- 15:38:54.260 [main] WARN  com.solac.messa.recei.RequestReplyMessageReceiverImpl - Application code threw an unhandled exception while processing request message in RequestMessageHandler
com.solace.messaging.PubSubPlusClientException$IncompatibleMessageException: Incompatible message, provided converter can't be used
        at com.solace.messaging.receiver.MessageReceiver$InboundMessageImpl.getAndConvertPayload(MessageReceiver.java:289) ~[solace-messaging-client-1.1.0.jar:?]
        at com.solace.samples.java.patterns.DirectReplierBlocking.lambda$main$2(DirectReplierBlocking.java:71) ~[solace-samples-java.jar:]
        at com.solace.messaging.receiver.RequestReplyMessageReceiverImpl.doRequestReply(RequestReplyMessageReceiverImpl.java:497) ~[solace-messaging-client-1.1.0.jar:?]
        at com.solace.messaging.receiver.RequestReplyMessageReceiverImpl.receiveMessage(RequestReplyMessageReceiverImpl.java:461) ~[solace-messaging-client-1.1.0.jar:?]
        at com.solace.samples.java.patterns.DirectReplierBlocking.main(DirectReplierBlocking.java:95) ~[solace-samples-java.jar:]
```
- I'm assuming it has to do with this code block?  I think this converter method is when you send a binary message and want to decode/deserialize it, as a DTO (Data Transfer Object).  But we are publishing Text messages, not binary messages containing UTF-8 strings.  So you can't just convert it.  If you use SdkPerf again, or dump(), you can see there are extra bytes, this is b/c TextMessages are a type of StructuredDataType.  I'm not sure how it's done in the New Java API, but there must be another way to just receive a TextMessage, and not use a converter. .... I just checked the JavaDocs, I think you want `getPayloadAsString()`.  https://docs.solace.com/API-Developer-Online-Ref-Documentation/pubsubplus-java/com/solace/messaging/Message.html#getPayloadAsString--  
```
            final String stringPayload = inboundMessage.getAndConvertPayload(
                    (byte[] bytes) -> {
                        return new String(bytes);
                    }, String.class
            );
```
- Get rid of received message rate, who cares since it's only 1 msg/s.  Also, the counter is never reset so it just keeps incrementing?  Also get rid of discard detection, only necessary on high-performance Direct apps, which these samples are not for.  Lines 63-67, 96-100, and 130-140.



That's it for now.  Thanks!

Commit:
Changes for fixing review comments

HariRangarajan-Solace commented:
> Just finished review. I previously suggested looking at JCSMP req/reply samples to see how they did things, and would recommend again. They are different than the pub/sub samples, topics, timeouts, functionality exposed, etc. Additionally, interoperability between the two should be tested... we are striving to be able to mix-and-match any API with any other. For example: the topics do not line up here between JCSMP and Java: `solace/samples/jcsmp/direct/request` vs. `solace/samples/java/direct/request/43`. Please test Java<->Java, JCSMP<->Java, and Java<->JCSMP.
> 
> Ok, comments:
> 
> **Requestor:**
> 
> * Filename, class name, and run scripts should be "requestOr", not "requestEr"
> * `REQUEST_TIMEOUT_MS = 10000;` seems a bit long?  Default in JCSMP and others is 3000
> * I'm still on the fence about these "helper" methods that only have a few lines in them, since you have to jump around in the code to see what they're doing.  Also, this is one is just 1 line, so why not have in the main body?
> 
> ```
>     private static String createMessagePayload(final int loopCounter) {
>         // each loop, change the payload, less trivial
>         return String.format("Hello, this is the request number:  #%d", loopCounter);    <--- extra space too
>     }
> ```
> 
> * The ApplicationMessageID is still not an actual app msg ID, it's a User Property.  But this is an API problem, not your code.  I'll ping the R&D guys about it.  I have the same issue on the DirectPublisher.  Here's the diff between Java API and JCSMP API:
> 
> ```
> JCSMP DirectPublisher:
> ^^^^^^^^^^^^^^^^^^ Start Message ^^^^^^^^^^^^^^^^^^^^^^^^^^^
> Destination:                            Topic 'solace/samples/jcsmp/direct/pub/V'
> ApplicationMessageId:                   790c72d5-2fc0-4a44-9211-e1789626687e
> SequenceNumber:                         74
> 
> Java DirectPublisher:
> ^^^^^^^^^^^^^^^^^^ Start Message ^^^^^^^^^^^^^^^^^^^^^^^^^^^
> Destination:                            Topic 'solace/samples/java/direct/pub/V'
> DMQ Eligible
> User Property Map:
>   Key 'solace.messaging.message.application-message-id' (STRING) 3d1eed99-5a03-44d4-ad91-5f26ea2c3372
> ```
> 
> * I'd suggest removing it from the Request/Reply samples as it's not super necessary here and already in the publisher.  And we're going to have to correct it once R&D fixes the API.
> * The Requestor errors out and quits if it doesn't receive a response in the time allotted.  Perhaps make it retry, loop, like JCSMP?         Oh!  I see from your comment above this was intentional.  Ok.  Well, I feel it should retry so that it's more resilient and doesn't matter which app (req or replier) is started first.
> 
> ```
>  } catch (JCSMPRequestTimeoutException e) {
>       System.out.println("Failed to receive a reply in " + REQUEST_TIMEOUT_MS + " msecs");
>   } catch (JCSMPException e) {  // threw from send(), only thing that is throwing here, but keep trying (unless shutdown?)
>       System.out.printf("### Caught while trying to producer.send(): %s%n", e);
> ```
> 
> * Means you have `catch` for exactly the right exception (line 87) since everything in Java API is a RuntimeException now.  Also, the error message on the console is incorrect: `Caught while trying to publisher.publish():` should be `publishAwaitResponse()`
> 
> **Replier:**
> 
> * Was this code run before submitting??  I'm getting unhandled exceptions the first time I try to run.
> 
> ```
> --LOG-- 15:38:54.260 [main] WARN  com.solac.messa.recei.RequestReplyMessageReceiverImpl - Application code threw an unhandled exception while processing request message in RequestMessageHandler
> com.solace.messaging.PubSubPlusClientException$IncompatibleMessageException: Incompatible message, provided converter can't be used
>         at com.solace.messaging.receiver.MessageReceiver$InboundMessageImpl.getAndConvertPayload(MessageReceiver.java:289) ~[solace-messaging-client-1.1.0.jar:?]
>         at com.solace.samples.java.patterns.DirectReplierBlocking.lambda$main$2(DirectReplierBlocking.java:71) ~[solace-samples-java.jar:]
>         at com.solace.messaging.receiver.RequestReplyMessageReceiverImpl.doRequestReply(RequestReplyMessageReceiverImpl.java:497) ~[solace-messaging-client-1.1.0.jar:?]
>         at com.solace.messaging.receiver.RequestReplyMessageReceiverImpl.receiveMessage(RequestReplyMessageReceiverImpl.java:461) ~[solace-messaging-client-1.1.0.jar:?]
>         at com.solace.samples.java.patterns.DirectReplierBlocking.main(DirectReplierBlocking.java:95) ~[solace-samples-java.jar:]
> ```
> 
> * I'm assuming it has to do with this code block?  I think this converter method is when you send a binary message and want to decode/deserialize it, as a DTO (Data Transfer Object).  But we are publishing Text messages, not binary messages containing UTF-8 strings.  So you can't just convert it.  If you use SdkPerf again, or dump(), you can see there are extra bytes, this is b/c TextMessages are a type of StructuredDataType.  I'm not sure how it's done in the New Java API, but there must be another way to just receive a TextMessage, and not use a converter. .... I just checked the JavaDocs, I think you want `getPayloadAsString()`.  https://docs.solace.com/API-Developer-Online-Ref-Documentation/pubsubplus-java/com/solace/messaging/Message.html#getPayloadAsString--
> 
> ```
>             final String stringPayload = inboundMessage.getAndConvertPayload(
>                     (byte[] bytes) -> {
>                         return new String(bytes);
>                     }, String.class
>             );
> ```
> 
> * Get rid of received message rate, who cares since it's only 1 msg/s.  Also, the counter is never reset so it just keeps incrementing?  Also get rid of discard detection, only necessary on high-performance Direct apps, which these samples are not for.  Lines 63-67, 96-100, and 130-140.
> 
> That's it for now. Thanks!

Hey @aaron-613 , I made the following changes as we discussed :

- Changed the file name for the request**o**r 😁 
- Updated the timeout duration
- Removed the application message properties as they are not used for the corelation 
- Put in a seperate catch for the TimeoutException where the flow wont break. Its possible to catch the RuntimeException seperately if we order the catches in the proper order. I also added in some comments which highlight the possibility for resilience implementation if required
- In the replier, I removed th explicit converter as the API is capable of identifying already what is the type of payload and converts it to a String appropriately.
- In the replier, I also removed the message counter, considering this is a blocking example its indeed redundant.
-
